numpy/_core/src/umath/ufunc_object.c

```diff
@@ -6635,6 +6635,57 @@ ufunc_set_doc(PyUFuncObject *ufunc, PyObject *doc, void *NPY_UNUSED(ignored))
     }
 }
 
+static PyObject *
+ufunc_get_inspect_signature(PyUFuncObject *ufunc, void *NPY_UNUSED(ignored))
+{
+    PyObject *signature;
+
+    // If there is a __signature__ in the instance __dict__, use it.
+    int result = PyDict_GetItemRef(ufunc->dict, npy_interned_str.__signature__,
+                                   &signature);
+    if (result == -1) {
+        return NULL;
+    }
+    else if (result == 1) {
+        return signature;
+    }
+
+    if (npy_cache_import_runtime(
+            "numpy._core._internal", "_ufunc_inspect_signature_builder",
+            &npy_runtime_imports._ufunc_inspect_signature_builder) == -1) {
+        return NULL;
+    }
+
+    signature = PyObject_CallFunctionObjArgs(
+            npy_runtime_imports._ufunc_inspect_signature_builder,
+            (PyObject *)ufunc, NULL);
+    if (signature == NULL) {
+        return NULL;
+    }
+
+    // Cache the result in the instance dict for next time
+    if (PyDict_SetItem(ufunc->dict, npy_interned_str.__signature__,
+                       signature) < 0) {
+        Py_DECREF(signature);
+        return NULL;
+    }
+
+    return signature;
+}
+
+static PyObject *
+ufunc_getattro(PyObject *obj, PyObject *name)
+{
+    // __signature__ special-casing to prevent class attribute access
+    if (PyUnicode_Check(name) &&
+        PyUnicode_CompareWithASCIIString(name, "__signature__") == 0) {
```

Comment:
Maybe compare with `npy_interned_str.__signature__`?

Aside: interesting that it needs the attribute getter, I had not thought of this... is it needed just for the caching in `__dict__`? since signature access is unlikely to be frequent or performance critical, I think it is actually fine not to cache. But this is OK too!